core/domain/role_services.py

```diff
@@ -131,3 +208,98 @@ def get_all_actions(role):
         role_actions.extend(get_all_actions(parent_role))
 
     return list(set(role_actions))
+
+
+def get_role_graph_data():
+    """Returns dict for displaying roles graph.
+
+    Returns:
+        dict. A dict containing data in following format:
+        {
+            links: list(dict(str:str)). List of dicts defing each edge in
+                following format:
+                    {
+                        source: Role Id from which edge is going out.
+                        target: Role Id to which edge is incoming.
+                    }
+            nodes: dict(str:str). Mapping of role Id to its human readable
+                format.
+        }
+    """
+    role_graph = {}
+    role_graph['links'] = []
+    role_graph['nodes'] = {}
+    for role in PARENT_ROLES:
+        role_graph['nodes'][role] = HUMAN_READABLE_ROLES[role]
+        for parent in PARENT_ROLES[role]:
+            role_graph['links'].append({'source': parent, 'target': role})
+    return role_graph
+
+
+def log_role_query(user_id, intent, role=None, username=None):
+    """Stores the query to role structure in RoleQueryAuditModel."""
+    model_id = '%s.%s.%s.%s' % (
+        user_id, int(math.floor(time.time())), intent, random.randint(0, 1000)
+    )
+
+    audit_models.RoleQueryAuditModel(
+        id=model_id, user_id=user_id, intent=intent,
+        role=role, username=username).put()
+
+
+def get_max_priority_role(role_list):
+    """Returns the role with maximum priority among the given list
+    of roles.
+
+    Args:
+        role_list: list(str). List of roles.
+
+    Returns:
+        str. Role with highest priority among given roles.
+    """
+    priority_num = -1
+    resultant_role = None
+    for role in role_list:
+        if PRIORITY_NUMBER_DICT[role] > priority_num:
+            resultant_role = role
+            priority_num = PRIORITY_NUMBER_DICT[role]
+    return resultant_role
+
+
+def get_role_changes(old_config_properties, new_config_properties):
+    """This function takes old and new versions of config property values
+    and returns the resultant roles for users in terms of new system.
+
+    Args:
+        old_config_properties: dict(str:list). Dict mapping config property ids
+            to their values.
+        new_config_properties: dict(str:list). Dict mapping config property ids
+            to their values.
+
+    Returns:
+        dict(str:str). Dict mapping usernames to roles in terms of new system.
+    """
+    changed_user_roles = {}
+    resultant_changed_roles = {}
+
+    for new_name, new_value in new_config_properties.iteritems():
```

Comment:
I'm a bit confused about all the for loops and if conditions here. Why not get the list of config property ids that are relevant and loop over just that list?